src/function/gds/shortest_paths.cpp

```diff
@@ -19,6 +19,50 @@ using namespace kuzu::graph;
 namespace kuzu {
 namespace function {
 
+enum class RecJoinOutputType : uint8_t {
+    DESTINATION_NODES = 0,
+    LENGTHS = 1,
+    // PATHS returns only the intermediate nodeIDs in a path, not the source or dst nodeIDs.
+    PATHS = 2,
+};
+
+/**
+ * A data structure to keep track of a single (shortest) path from one source node to
+ * multiple destination (and intermediate) nodes as "backward BFS graph" form. The data structure is
+ * "dense" in the sense that it keeps space for one "backwards edge" for each possible node that can
+ * be in the destination. Supports multi-label nodes.
+ */
+class SinglePaths {
+public:
+    explicit SinglePaths(
+        std::vector<std::tuple<common::table_id_t, uint64_t>> nodeTableIDAndNumNodes,
+        MemoryManager* mm) {
+        for (const auto& [tableID, numNodes] : nodeTableIDAndNumNodes) {
+            lastEdges.insert({tableID, mm->allocateBuffer(false, numNodes * sizeof(nodeID_t))});
```

Comment:
this is fine since we are still prototyping. Though I doubt we will have good performance if we always allocate towards maxNumNodes for each source. In my mind a dynamic data structure is more robust than this static array.

Reply:
I think dense structures will be more robust for most cases. I have a todo "Experiment with sparse-vs-dense implementation switching" here: https://github.com/kuzudb/kuzu/issues/3717, which I will get to at some point.